src/Groups/GAPGroups.jl

```diff
@@ -0,0 +1,617 @@
+# further possible functions: similar, literal_pow, parent_type
+
+export order, perm, cperm, isfinite, gens, ngens, comm, comm!, inv!, rand_pseudo, one!, div_right,
+div_left, div_right!, div_left!, deg, mul, mul!, listperm, degree, elements, right_coset, coset_decomposition,
+right_cosets , right_transversal, conjugacy_class, conjugacy_classes, number_conjugacy_classes, isconjugate, conjugacy_classes_subgroups, conjugacy_classes_maximal_subgroups, normalizer, normaliser, core, pcore, fitting_subgroup, frattini_subgroup, radical_subgroup, socle, sylow_subgroup, hall_subgroup, sylow_system, complement_system, hall_system, relators, isperfect, issimple, isalmostsimple, ispgroup
+
+
+# TODO: as soon as GAP packages like `polycyclic` or `rcwa` are loaded,
+# the custom group types and isos they define should be added to the arrays
+# _gap_group_types resp. _iso_function
+
+
+function group_element(G::T, x::GapObj) where T <: Group
+  return GAPGroupElem{T}(G, x)
+end
+
+function elements(G::T) where T <: Group
+  els = GAP.gap_to_julia(GAP.Globals.Elements(G.X))
+  elems = Vector{elem_type(G)}(undef, length(els))
+  i = 1
+  for x in els
+    elems[i] = group_element(G, x)
+    i += 1
+  end
+  return elems
+end
+
+# to be fixed later
+Base.hash(x::GAPGroupElem) = 0
+
+function parent(x::GAPGroupElem)
+  return x.parent
+end
+
+function Base.isfinite(G::PermGroup)
+  return true
+end
+
+function Base.isfinite(G::PcGroup)
+  return true
+end
+
+function Base.isfinite(G::Group)
+  return GAP.Globals.IsFinite(G.X)
+end
+
+function degree(x::PermGroup)
+   return x.deg
+end
+
+function order(x::Group)
+   return GAP.gap_to_julia(GAP.Globals.Size(x.X))
+end
+
+function order(x::GAPGroupElem)
+   return GAP.gap_to_julia(GAP.Globals.Order(x.X))
+end
+
+function order(::Type{T}, x::Union{GAPGroupElem, Group}) where T<:Number
+   return T(order(x))
+end
+
+Base.exponent(x::Group) = GAP.Globals.Exponent(x.X)
+
+Base.length(x::Group) = order(x)
+
+function Base.rand(x::Group)
+   s = GAP.Globals.Random(x.X)
+   return group_element(x, s)
+end
+
+# one of the following should be non-parametric
+rand_pseudo(G::Group) = Base.rand(G)
+
+
+function _maxgroup(x::T, y::T) where T <: Group
+   if x == y
+     return x
+   elseif GAP.Globals.IsSubset(x.X, y.X)
+     return x
+   elseif GAP.Globals.IsSubset(y.X, x.X)
+     return y
+   else
+     error("Not yet implemented")
+   end
+end
+
+#We need a lattice of groups to implement this properly
+function _prod(x::T, y::T) where T <: GAPGroupElem
+  G = _maxgroup(parent(x), parent(y))
+  return group_element(G, x.X*y.X)
+end
+
+Base.:*(x::GAPGroupElem, y::GAPGroupElem) = _prod(x, y)
+
+function ==(x::Group, y::Group)
+   return x.X == y.X
```

Comment:
Hm, is that right? Shouldn't it also e.g. take the degree into account for PermGroup?

Reply:
At the moment, when any group / subgroup is defined, the degree is assigned by default as the largest moved point. For example, if I type

> `G=symmetric_group(4)`

> `H = sub(G, gens(symmetric_group(3)) )[1]`,
it does not distinguish H by Sym(3), because the

Reply:
The subgroup *must* have the same degree as its overgroup! Otherwise the whole idea of tracking the parent implicitly as `sym(deg)` breaks down

Reply:
Ok. I corrected the two functions. Now `H = sub(G, gens(symmetric_group(3)) )[1]` has degree 4 instead of 3. Now, a similar problem rises: if `f = isisomorphic(H, symmetric_group(3)) [2]` is the isomorphism between `H` and Sym(3), then `f == id_hom(H)` or `f == id_hom(symmetric_group(3))` returns `true` (in other words, the function `isisomorphism` does not realize that `f` is a morphism between different groups. Should this also be fixed?

Reply:
Related question. If `G = symmetric_group(4)`, `H = sub(G, gens(symmetric_group(3)))` and `x in H`, do we want also that `x in symmetric_group(3)` returns `false` ? At the moment, it returns `true` since it relies on the GAP instructions, that does not distinguish parent of permutations.